packages/react-reconciler/src/ReactFiberThenable.new.js

```diff
@@ -110,24 +110,36 @@ export function trackUsedThenable<T>(thenable: Thenable<T>, index: number): T {
         // it's defined, but an unknown value, assume it's been instrumented by
         // some custom userspace implementation. We treat it as "pending".
       } else {
-        const pendingThenable: PendingThenable<mixed> = (thenable: any);
+        const pendingThenable: PendingThenable<T> = (thenable: any);
         pendingThenable.status = 'pending';
         pendingThenable.then(
           fulfilledValue => {
             if (thenable.status === 'pending') {
-              const fulfilledThenable: FulfilledThenable<mixed> = (thenable: any);
+              const fulfilledThenable: FulfilledThenable<T> = (thenable: any);
               fulfilledThenable.status = 'fulfilled';
               fulfilledThenable.value = fulfilledValue;
             }
           },
           (error: mixed) => {
             if (thenable.status === 'pending') {
-              const rejectedThenable: RejectedThenable<mixed> = (thenable: any);
+              const rejectedThenable: RejectedThenable<T> = (thenable: any);
               rejectedThenable.status = 'rejected';
               rejectedThenable.reason = error;
             }
           },
         );
+
+        // Check one more time in case the thenable resolved synchronously
+        switch (thenable.status) {
```

Comment:
@acdlite this never gets hit, since this always results in `'pending'`:
```js
const syncThen = (p) => {
  let status = 'pending'
  p.then(() => status = 'fulfilled')
  console.log(status)
}
syncThen(Promise.resolve()) // 'pending'
syncThen(new Promise(r => r())) // 'pending'
const resolved = new Promise(r => r())
syncThen(resolved) // 'pending'
```
I believe it's because `.then` always gets executed asynchronously, like `setTimeout(..., 0)`